orangecontrib/text/preprocess.py

```diff
@@ -1,94 +1,269 @@
+            pp_corpus = self._preprocess_document(data)
+            return pp_corpus
+        elif isinstance(data, list):
+            if self.use_df_sw:  # DF parameters were specified.
+                self._get_corpus_specific_stopwords(data)
+            pp_corpus = [self._preprocess_document(doc) for doc in data]
+            return pp_corpus
+        else:
+            raise ValueError('Type "{}" not supported.'.format(type(data)))
+
+    def _preprocess_document(self, data):
+        # Tokenize.
+        data = self._tokenize(data)
+        # Lowercase.
+        if self.lowercase:
+            data = [token.lower() for token in data]
+        # Remove stop words.
+        if self.stop_words is not None:
+            data = self._remove_stop_words(data, self.stop_words)
+        if self.df_stop_words is not None:
+            data = self._remove_stop_words(data, self.df_stop_words)
+        # Transform.
+        if self.transformation is not None:
+            data = self._transform(data)
+        return data
+
+    def _tokenize(self, data):
+        """Splits the text of the input into tokens by whitespaces and punctuation.
+
+        Args:
+            data (str): The input holding the text to be tokenized.
+
+        Returns:
+            list: Pre-processed document in the form of a list, holding the tokens.
+        """
         if isinstance(data, str):
-            vec = self.cv.fit([data])
-            return vec.get_feature_names()
-        if isinstance(data, list):
-            vec = self.cv.fit(data)
-            features = vec.get_feature_names()
-            docs = [[] for _ in range(len(data))]
-            for (line, column), count in vec.transform(data).todok().items():
-                docs[line].extend([features[column]] * count)
-            return docs
+            return self.tokenizer(data)
         else:
-            raise ValueError("Type '{}' not supported.".format(type(data)))
+            raise ValueError('Type {} not supported.'.format(type(data)))
+
+    def _transform(self, data):
+        """Performs on the input, whatever transformation was specified at class instance creation.
+
+        Args:
+            data (str, list): The input holding the text to be transformed. Can be provided
+                in form of a string or list of string tokens.
+
+        Returns:
+            str, list: Depending on the input, returns either a list of tokens or a single token.
+        """
+        return self.transformation(data)
+
+    @staticmethod
+    def _remove_stop_words(data, stop_words):
+        """Removes the stop words from the corpus.
+
+        Removes the stop words specified in the stop_words parameter.
+
+        Args:
+            data (list): The input document in the form of a list.
+
+        Returns:
+            list: Pre-processed document in the form of a list, holding the tokens.
+        """
+        return [word for word in data if word not in stop_words]
```

Comment:
Love your Napoleon, thanks.

Could you please add above this line the following:
```py
assert isinstance(stop_words, set)
```
We need to ensure `stop_words` provides fast lookup. It's O(n*m) now whereas it could be O(n).